### PR TITLE
Converted to async code using async await.

### PR DESCRIPTION
**Pull request by sejonssonr**
Converted code to non blocking using async await. Also fixed a bug causing requests to request too much data.

### CODE REVIEW

File: IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/BaseModbusSlaveConfig.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ public class BaseModbusSlaveConfig
         [DefaultValue(50)]
         [Range(1, int.MaxValue)]
         [JsonProperty(Required = Required.Default, DefaultValueHandling = DefaultValueHandling.Populate)]
-        public int? RetryInterval { get; set; }
+        public int RetryInterval { get; set; }
```
Comment (danielado) [inline IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/BaseModbusSlaveConfig.cs:25]: Could the type be changed from int? to int for RetryInterval and TcpPort as well?

File: IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/Decoding/ValueDecoderBase.cs [hunk 2]
```diff
@@ -30,7 +30,10 @@ public IEnumerable<DecodedValue> GetValues(byte[] bytesToConvert, ReadOperation
             return result;
         }
 
-        public virtual short GetByteCount(short valuesToRead) => checked((short)(this.ByteSize * valuesToRead));
+        /// <summary>
+        /// Get the number of 16 bit registers that should be read for the specified number of values to read. 
+        /// </summary>
+        public virtual short GetRegisterCount(short valuesToRead) => checked((short)(this.ByteSize / 2 * valuesToRead));
```
Comment (danielado) [inline IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/Decoding/ValueDecoderBase.cs:36]: This implies that ByteSize >= 2 for any ValueDecoderBase. Maybe add a comment?

**Review by suneetnangia — APPROVED** [review]
Seems reasonable and urgent fixes hence progressing immediately.

File: IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusRTUSlaveSession.cs [hunk 3]
```diff
@@ -68,7 +68,7 @@ protected override void EncodeRead(ReadOperation operation)
             operation.Request[this.FunctionCodeOffset + 1] = addressBytes[0];
             operation.Request[this.FunctionCodeOffset + 2] = addressBytes[1];
             //count
-            byte[] countBytes = BitConverter.GetBytes(IPAddress.HostToNetworkOrder(operation.Decoder.GetByteCount(operation.Count)));
+            byte[] countBytes = BitConverter.GetBytes(IPAddress.HostToNetworkOrder(operation.Decoder.GetRegisterCount(operation.Count)));
```
Comment (danielado) [inline IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusRTUSlaveSession.cs:71]: Use the same name for countBytes as in ModbusTCPSlaveSession.

File: IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/SessionsHandle.cs [hunk 4]
```diff
@@ -48,18 +48,18 @@ public async Task<SessionsHandle> CreateHandleFromConfiguration(ModuleConfig con
             return sessionsHandle;
         }
 
-        public List<ModbusSlaveSession> ModbusSessionList = new List<ModbusSlaveSession>();
+        public readonly List<ModbusSlaveSession> ModbusSessionList = new List<ModbusSlaveSession>();
 
         public void Release()
```
Comment (danielado) [inline IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/SessionsHandle.cs:53]: This method should also be made async. Then await all session.ReleaseSessionAsync calls.

File: IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusTCPSlaveSession.cs [hunk 5]
```diff
@@ -1,247 +1,204 @@
...
+            else
+            {
+                byte[] valueBytes = BitConverter.GetBytes(IPAddress.HostToNetworkOrder((Int16)valueToWrite));
+                request[this.FunctionCodeOffset + 3] = valueBytes[0];
+                request[this.FunctionCodeOffset + 4] = valueBytes[1];
+            }
+        }
+        protected override async Task<byte[]> SendRequestAsync(byte[] request)
+        {
+            byte[] response = null;
```
Comment (danielado) [inline IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusTCPSlaveSession.cs:132]: Use var for the variables here.

File: IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusTCPSlaveSession.cs [hunk 6]
```diff
@@ -1,247 +1,204 @@
...
+            }
+
+            return response;
+        }
+
+        private async Task<byte[]> ReadResponseAsync()
+        {
+            byte[] headerResponse = await this.ReadMBAPHeaderResponseAsync().ConfigureAwait(false);
+
+            var mbapHeaderByteCountStartIndex = 4;
```
Comment (danielado) [inline IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusTCPSlaveSession.cs:179]: var -> const int

File: IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusTCPSlaveSession.cs [hunk 7]
```diff
@@ -1,247 +1,204 @@
...
+            var bytesToRead = IPAddress.NetworkToHostOrder((Int16)BitConverter.ToUInt16(headerResponse, mbapHeaderByteCountStartIndex)) - 1;
```
Comment (danielado) [inline IoTEdgeModbus/Modules/AzureIoTEdgeModbus/Slave/ModbusTCPSlaveSession.cs:180]: Could add a constant for the - 1 magic number.

**Pull request merged**